### PR TITLE
Dataportal Client And Metadata Adapter

### PR DESCRIPTION
**Pull request by mbruns91**
This pull request adds the public Dataportal client and its service-specific metadata adapter. Generic publication information remains owned by `PublicationMetadata`, while `DataportalMetadata` contains only Dataportal and CKAN concerns and converts both layers into a CKAN package payload.

## Summary

- Added `DataportalClient` with the default `dataportal.material-digital.de` address and support for custom deployments.
- Exported `DataportalClient` and `DataportalMetadata` from `courier.services.dataportal`, but not from the top-level package.
- Required generic fields to enter through `PublicationMetadata`.
- Restricted `DataportalMetadata` to Dataportal/CKAN fields such as `name`, `owner_org`, `private`, `groups`, `extras`, and `dataset_type`.
- Mapped publication fields into CKAN fields and deterministic structured extras.
- Added package-name derivation and metadata collision validation.
- Commits: `64b5618 Add Dataportal client skeleton`, `82b6062 Add Dataportal metadata adapter`, and `1d7a0ab Test Dataportal metadata boundary`.

## Example

```python
from courier.metadata import Person, PublicationMetadata
from courier.services.dataportal import DataportalClient, DataportalMetadata

client = DataportalClient(api_token="secret")
publication = PublicationMetadata(
    title="Example dataset",
    description="Example description",
    creators=(Person(name="Example Author"),),
)
metadata = DataportalMetadata(
    metadata=publication,
    owner_org="example-organization",
)
dataset = client.packages.create(metadata.to_payload())
```

## Unit tests

- Default and custom Dataportal addresses.
- Public export boundaries.
- Publication-to-CKAN field conversion.
- Deterministic structured-extra serialization.
- Name derivation and blank-field validation.
- Rejection of generic publication fields passed directly to `DataportalMetadata`.

## Validation

- `black`
- `ruff check`
- `pytest`
- `mypy

### CODE REVIEW

File: courier/services/dataportal/metadata.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            _ = _required_string(self.dataset_type, "dataset_type")
+        if self.private is not None and not isinstance(self.private, bool):
+            raise ValidationError("private must be a boolean when provided")
+
+        for group in self.groups:
+            _ = _required_string(group, "group")
+
+        normalized_extra_keys: set[str] = set()
+        for key, value in self.extras.items():
+            normalized_key = _required_string(key, "extra key")
```
[the remark this answers is not in the document]
Comment (mbruns91) [reply courier/services/dataportal/metadata.py:53]: Fixed via [1995274](https://github.com/pyiron/courier/pull/74/commits/199527402bcf08d034e31c3657785a0083ad7964). `DataportalMetadata.validate()` now explicitly verifies that `groups` is a list and `extras` is a dictionary before using either container. Invalid values such as `None`, strings for `groups`, and lists for `extras` now consistently raise `ValidationError`. Regression tests cover these cases.

File: courier/services/dataportal/metadata.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    """Adapt service-neutral publication metadata to a CKAN package payload."""
+
+    metadata: PublicationMetadata
+    name: str | None = None
+    owner_org: str | None = None
+    private: bool | None = None
+    groups: list[str] = field(default_factory=list)
+    extras: dict[str, str] = field(default_factory=dict)
+    dataset_type: str | None = None
+
```
Comment (mbruns91) [inline courier/services/dataportal/metadata.py:26]: @samwaseda I think that's the only interesting part for you.
